raidemulator: Make websocket connection optional

valarnin opened the pull request:
This has been possible for a while, since the switch to webpack for manifest files, I just never got around to making websocket optional.

The change to Persistor is due to an unintentional race condition with event ordering which was only apparent when there was an actual delay to websocket connection.

It might make sense additionally to check for OVERLAY_WS in the URL and skip the attempt to connect entirely?

Trim21 commented:
Could user triggers work without connection to ACT?

valarnin commented:
It's fully functional. Importing logs and running through encounters. You'll need the fixes to getCombatants if you want to test against e12s p1, but otherwise everything should work except user config changes and triggers.

Trim21 commented:
> It's fully functional. Importing logs and running through encounters. You'll need the fixes to getCombatants if you want to test against e12s p1, but otherwise everything should work except user config changes and triggers.

user config is not loaded

Trim21 commented:
I fixed getCombatants and find that missing many triggers offline

valarnin commented:
> but otherwise everything should work **_except user config changes and triggers_**

Also, the updated intro text reflects this:

![image](https://user-images.githubusercontent.com/6119598/119849100-56013600-beda-11eb-8cd6-57b49e47f3a1.png)

Trim21 commented:
sorry, I misunderstand `fully functional`

Trim21 commented:
Test on E9S, looks like not working correctly.

![image](https://user-images.githubusercontent.com/13553903/119851966-76b0a680-bf41-11eb-84c1-b6de4147bb2b.png)

offline

![image](https://user-images.githubusercontent.com/13553903/119852047-87f9b300-bf41-11eb-8a44-9dd28b5e5aa6.png)

valarnin commented:
To narrow this down, was this a full test for both online and offline (i.e. from an empty DB)? Or were these logs already loaded before testing?

Trim21 commented:
pre-existed DB before I checkout PR

valarnin commented:
Hmm, I can't reproduce this though I did spot another issue with the `persistor.on('ready'` call. I don't have any e9s logs saved locally right now, can you check if this happens with other fights, e.g. e11s/e12s/dia ex?

valarnin commented:
I decided to change this to a draft PR as there are still some other outstanding issues with this adjustment, related to the timeline display.

Trim21 commented:
I want to export my db and upload here, but got a out of memory problems.

valarnin commented:
All the raidemulator stuff is stored in IndexedDB, you can just manually dump the one encounter and I can manually import on my side.

valarnin commented:
Rebased to upstream/main to pick up the relative offsets PR's changes, and confirmed that fixes the issues I was spotting. Also fixed two other issues. @Trim21 Any luck with getting an export of that encounter for me to try and debug?

Trim21 commented:
any js snippets to export indexeddb?

valarnin commented:
Hmm, double checking your screenshots, it's possible this is due to language differences? In raidemulator.js, can you adjust as follows:

```javascript
import Options from './raidboss_options';
import { callOverlayHandler } from '../../resources/overlay_plugin_api';

console.log(
    Options.ParserLanguage,
    Options.ShortLocale,
    Options.DisplayLanguage);

/*Options.ParserLanguage = 'en';
Options.ShortLocale = 'en';
Options.ShortLocale = 'en';*/

(() => {
```

Check the console logged values when connected to websocket, then try manually forcing them to be correct when not connected to websocket?

Trim21 commented:
they are the same

Trim21 commented:
And how do you workaround the bug of `ttsSay` in emulator...

valarnin commented:
`ttsSay` is a separate issue that I still need to look into. It should be falling back to browser speechSynthesis automatically but isn't.

So you have `en en en` in your console when connected to websocket?

![image](https://user-images.githubusercontent.com/6119598/120115575-4d1a9980-c152-11eb-9a63-fa05b525682f.png)

Trim21 commented:
yes, with or without ws, `en en en`

valarnin commented:
Oh right, console logging there is going to show the defaults, not the websocket-fetched values...

Move the console.log down such that:

```javascript
        if (websocketConnected) {
          await UserConfig.getUserConfigLocation('raidboss', Options, (e) => {
            document.querySelector('.websocketConnected').classList.remove('d-none');
            document.querySelector('.websocketDisconnected').classList.add('d-none');
          });
          console.log(
              Options.ParserLanguage,
              Options.ShortLocale,
              Options.DisplayLanguage);
        }
```

Trim21 commented:
still

valarnin commented:
Can you confirm what your actual settings are in ACT?
![image](https://user-images.githubusercontent.com/6119598/120116706-6c67f580-c157-11eb-9319-0af25c512056.png)
![image](https://user-images.githubusercontent.com/6119598/120116713-74279a00-c157-11eb-85b5-13774c684827.png)

Trim21 commented:
they&nbsp;are&nbsp;all CN



------------------&nbsp;Original&nbsp;------------------
From: ***@***.***&gt;; 
Date: 2021年5月31日(星期一) 凌晨2:58
To: ***@***.***&gt;; 
Cc: ***@***.***&gt;; ***@***.***&gt;; 
Subject: Re: [quisquous/cactbot] raidemulator: Make websocket connection optional (#2942)





 
Can you confirm what your actual settings are in ACT?
 
 
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

valarnin commented:
Try overriding those three `Options` params to be `'cn'` and see if your e9s encounter populates correctly?

Trim21 commented:
I finnally dumped the database with jsonlines format.

But it's hard to share 
![image](https://user-images.githubusercontent.com/13553903/120117560-52163200-c1c0-11eb-87de-41261b357b09.png)

Trim21 commented:
uploaded here https://github.com/Trim21/actions-cron/releases/tag/1



```js
  async exportDB() {
    // streamSaver.createWriteStream() returns a writable byte stream
    // The WritableStream only accepts Uint8Array chunks
    // (no other typed arrays, arrayBuffers or strings are allowed)
    if (started)
      return;
    started = true;
    const fileStream = streamSaver.createWriteStream('db.jsonlines', {
      // size: uInt8.byteLength, // (optional filesize) Will show progress
      writableStrategy: true, // (optional)
      readableStrategy: true, // (optional)
    });

    // using Response can be a great tool to convert
    // mostly anything (blob, string, buffers) into a byte stream
    // that can be piped to StreamSaver
    //
    // You could also use a transform stream that would sit
    // between and convert everything to Uint8Arrays
    // new Response('StreamSaver is awesome').body
    //   .pipeTo(fileStream)
    //   .then(success, error);

    const ret = {
      encounters: [],
    };
    const summaries = await this.listEncounters();
    const writer = fileStream.getWriter();
    let count = 0;
    const encode = TextEncoder.prototype.encode.bind(new TextEncoder);
    for (const summary of summaries) {
      const enc = await this.loadEncounter(summary.id);
      const line = {
        encounterDay: EmulatorCommon.timeToDateString(summary.Start),
        encounterZoneName: summary.ZoneName,
        encounterZoneId: summary.ZoneId,
        encounterLines: enc.logLines,
      };
      count++;
      await writer.write(encode(JSON.stringify(line) + '\n'));
      // ret.encounters.push({
      //   encounterDay: EmulatorCommon.timeToDateString(summary.Start),
      //   encounterZoneName: summary.ZoneName,
      //   encounterZoneId: summary.ZoneId,
      //   encounterLines: enc.logLines,
      // });
      console.log(count);
    }
    console.log('end');
    console.log(count, writer.size);
    await writer.close();
    started = false;
    return ret;
  }
```

if `ret.encounters` in jsonlines format.

valarnin commented:
I have confirmed based on your exported encounters that this is 100% related to the `Options.ParserLanguage` setting. Forcing this to `cn` allows the logs to parse fine (conversion on import works without this being set, that part is language-agnostic).

Right now, locally, I have a few additional features to address this issue.

1. If you load the page with a websocket at any point, it'll cache the responses for `cactbotLoadUser` and `cactbotLoadData`. If you then load without a websocket in the future, it'll use those cached values. Cached values are overwritten any time those calls succeed.
2. If you load without a cached value, it will run the following code:
```javascript
          const browserLang = navigator.language.split('-')[0];
          Options.ParserLanguage = browserLang in Lang ? browserLang : 'en';
          // Default options
          Options.IsRemoteRaidboss = true;
          Options.TextAlertsEnabled = true;
          Options.SoundAlertsEnabled = true;
          Options.SpokenAlertsEnabled = false;
          Options.GroupSpokenAlertsEnabled = false;
```

I'm not sure if this is the best way to go about determining the correct language.

The other possibility, and this one probably makes the most sense, is to look at `Encounter.language` and basically override the global Options setting any time an encounter is loaded.

Thoughts?

quisquous commented:
What about the `249|2021-05-28T11:29:32.5221700-07:00|Language ID: 1, Disable Damage Shield: False, Disable Combine Pets: False, Parse Filter: None, DoTCrits: False, RealDoTs: False, Parse Potency: False|f29843bdf63dc293e025b5bcb04b9a64` line?

valarnin commented:
The `249|` line is only printed once on ACT start.

I guess this could also be tracked in NetworkLogConverter, and recreated for each encounter? It's not going to help for split logs or anything though, I believe?

Trim21 commented:
try to extract boss names？

valarnin commented:
It's difficult to determine which enemy is the boss programatically, but that logic is already in raidemulator (that's how it determines the primary combatant for displaying in the bottom tab view). However, that doesn't directly lead to which language is currently in use.

I guess we could look at the code points in unicode and try to match them up with JP/CN/KO but that won't really help for DE/FR/EN langs...

But all of this is kind of moot because there's already language autodetection based on the countdown/seal/engage/unseal regexes (see EmulatorCommon's matchStart/matchEnd/doesLineMatch functions), and those are also the lines that are used to determine valid encounters when doing the initial log import and parse.

The question here is, does it make sense to manually set the `Options.ParserLanguage` setting based on the encounter's detected language, given that not doing so will 100% cause PopupText's trigger matching to fail?

quisquous commented:
I think if it can be done automatically, then all the better.

> The `249|` line is only printed once on ACT start.
> 
> I guess this could also be tracked in NetworkLogConverter, and recreated for each encounter? It's not going to help for split logs or anything though, I believe?

It's true it's printed only at the start, but is that an insurmountable issue? Could the encounter store the language it finds, and the converter can set that as it finds encounters?

It looks like the cactbot log splitter already includes these lines in any split encounter.  (You can see this in the o12s log I attached a while back).  Are there other splitters people use?

valarnin commented:
Functionally we're talking about three different parts of data flow here.

1. Log import
   - Loops over a given file
   - Splits that file into encounters
   - Parses each line, extracting information and creating a converted line for use by timelines
   - Saves encounters that have actual content to IndexedDB
2. Encounter Analysis
   - Takes a parsed encounter
   - Analyzes it for executed triggers via extended PopupTextAnalysis and other classes
   - Prepares the analyzed encounter for playback
3. Encounter Playback
   - Play back encounter in real time
   - Active timeline/trigger display

Functionally, the best place to store off the encounter language is `#1`. It's relatively trivial to take the determined language from matchStart/matchEnd and save that off in the encounter, and then when loading (`#2`) or setting the currently active encounter (`#3`), also set the Options.ParserLanguage setting appropriately.

quisquous commented:
Yeah, sorry for not being precise.  That was the data flow I was envisioning too, where you figure out the language on import and store it per encounter for later.

valarnin commented:
Flagging this as ready for review again, since all the relevant issues seem to be resolved now.

valarnin commented on ui/raidboss/emulator/overrides/RaidEmulatorOverlayApiHook.js:
```diff
@@ -20,47 +41,79 @@ export default class RaidEmulatorOverlayApiHook {
+      return await this._getCombatantsOverride(msg);
+    } else if (cachedCactbotCalls.includes(msg.call)) {
+      if (!this.connected) {
+        if (this.hasCachedCactbotCall(msg))
+          return await this._cactbotCachedCall(msg);
```
I was thinking of an icon top-right or something that shows connected/cached/disconnected, maybe with a hover for cached time. Or maybe if it's cached/disconnected, it should always show the modal, or show a different dialog?

quisquous replied:
I think an icon would be good, because then if people send screenshots of raidemulator not working we can see it.

valarnin replied:
Added a modal when opening raidemulator if disconnected, as well as connection indicator. The indicator has a tooltip as well.

![image](https://user-images.githubusercontent.com/6119598/121913777-ae6c6c00-ccff-11eb-99ba-d45ca132f869.png)
![image](https://user-images.githubusercontent.com/6119598/121913829-b9270100-ccff-11eb-95df-49f7e362a864.png)



![image](https://user-images.githubusercontent.com/6119598/121913903-ca700d80-ccff-11eb-8575-db2326e8c938.png)
![image](https://user-images.githubusercontent.com/6119598/121913950-d360df00-ccff-11eb-9ea4-7976897e89d1.png)



![image](https://user-images.githubusercontent.com/6119598/121914021-e5db1880-ccff-11eb-9464-586ad383ed53.png)

quisquous replied:
Is there like a plug with a line through it for no, or something to visually distinguish them other than red/green colors?

Also sorry, but I still continue to think that caching will lead to confusion here and it'd be better to only support connected/unconnected states and not also cached (at some point).

valarnin replied:
Cached is an addition with this PR, I can remove that if you're against it entirely.

The icons are just FontAwesome icons, I didn't see a good option for disconnected plug or anything. If you have a better suggestion for a connected and/or disconnected icon, it's easy enough to adjust.

https://fontawesome.com/v5.15/icons?d=gallery&p=2&m=free

quisquous replied:
Yeah, sorry.  I am against it a little bit, mostly because I feel like cactbot gets enough support as it is, so keeping modes simpler for users goes a long way towards keeping cactbot more explainable.  If you feel like this would be really useful, I'm happy to hear you out?

Re: icons.  I could imagine `handshake` vs `handshake-slash`.  Alternatively `plug` vs `ban`?

valarnin replied:
The biggest reason for caching is language settings, in my opinion. The rest are irrelevant.

quisquous replied:
Gotcha!  I think in that case, I think it'd be better to figure out the language from the log file itself, like along the lines of https://github.com/quisquous/cactbot/pull/2942#issuecomment-851104263.  It'd be extremely convenient to be able to just point a user at the gh-pages version of raidemulator to test out a log file without having any other dependencies.   Having caching wouldn't help that use case.

I don't think that has to be done in this patch, but mostly trying to say "that's how I think this could be solved in the future".  This keeps "what language am I parsing in" independent of any ambiguous previous state and dependent on the log file itself.

valarnin replied:
It already figures out parser language from the log file, the cached language settings are for trigger and timeline display. If caching is dropped and a CN user is using raidboss, it'll show triggers and timelines in chinese if they're connected to ACT and english if not.

quisquous replied:
Oh, I see.  The display language, not the parsing language.  We could default to browser or parser language as the display language in those cases then?

valarnin replied:
This javascript should detect the most appropriate language and fall back to 'en' by default, if someone with a non-'en' language could verify?
```js
[...navigator.languages, 'en']
  .map((l) => l.split('-')[0].toLowerCase())
  .filter((l) => ['en', 'de', 'fr', 'ja', 'cn', 'ko'].includes(l))[0];
```

valarnin replied:
Alright, these changes are made. Please note if you're testing these that the timeline trigger bug is still present in this PR, so test against an encounter without timeline triggers or you're goona have a bad time.

valarnin commented on ui/raidboss/raidemulator.html:
```diff
@@ -68,9 +68,10 @@ <h5 class="modal-title">Introduction</h5>
+            <p>If connected to a web socket, this tool will load and respect user configuration files for cactbot/raidboss.</p>
             <p>No overlays need to be created.</p>
-            <p>Current WebSocket status: <span class="d-none websocketConnected text-success">Connected</span><span class="websocketDisconnected text-error">Disconnected</span>.</p>
+            <p>Current WebSocket status: <span class="d-none websocketConnected text-success">Connected</span><span class="d-none websocketCached text-warning">Cached</span><span class="websocketDisconnected text-warning">Disconnected</span>.</p>
```
Yeah, there's a fair amount of text for emulator that I'd like to offer translation capability for, I was thinking of making a pass at that after doing all the typescript conversion.
